src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -494,18 +585,56 @@ int encryptFinal(byte[] in, int inOfs, int len, byte[] out, int outOfs)
             doLastBlock(in, inOfs, len, out, outOfs, true);
         }
 
-        byte[] lengthBlock =
-            getLengthBlock(sizeOfAAD, processed);
-        ghashAllToS.update(lengthBlock);
+        byte[] block = getLengthBlock(sizeOfAAD, processed);
+        ghashAllToS.update(block);
         byte[] s = ghashAllToS.digest();
-        byte[] sOut = new byte[s.length];
+        if (tagLenBytes > block.length) {
+            block = new byte[tagLenBytes];
+        }
         GCTR gctrForSToTag = new GCTR(embeddedCipher, this.preCounterBlock);
-        gctrForSToTag.doFinal(s, 0, s.length, sOut, 0);
+        gctrForSToTag.doFinal(s, 0, s.length, block, 0);
 
-        System.arraycopy(sOut, 0, out, (outOfs + len), tagLenBytes);
+        System.arraycopy(block, 0, out, (outOfs + len), tagLenBytes);
         return (len + tagLenBytes);
     }
 
+    int encryptFinal(ByteBuffer src, ByteBuffer dst)
+        throws IllegalBlockSizeException, ShortBufferException {
+        int len = src.remaining();
+        dst.mark();
+        if (len > MAX_BUF_SIZE - tagLenBytes) {
+            throw new ShortBufferException
+                ("Can't fit both data and tag into one buffer");
+        }
+
+        if (dst.remaining() < len + tagLenBytes) {
+            throw new ShortBufferException("Output buffer too small");
+        }
+
+        checkDataLength(processed, len);
+
+        processAAD();
+        if (len > 0) {
```

Comment:
Even if (len == 0), we should still process the data stored into 'ibuffer'? It seems that both of the encrypt(ByteBuffer) and encryptFinal(ByteBuffer) are adapted from their counterpart with byte[] arguments. However, the byte[] methods have different entrant conditions due to the buffering in CipherCore. So the impl of the ByteBuffer ones may need additional logic to handle all possible calling sequence.

Reply:
Yes for encryptFinal.  encrypt is removed